### PR TITLE
Fallback to delete resources without tagging client

### LINKED ISSUE #3987: ExternalResourceGC feature flag needs to fall back when ResourceTagging API is unavailable (e.g. 'airgapped installs')
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

AWS have not launched a VPC endpoint for the ResourceTagging API, as such the ExternalResourceGC does not work when there is no egress to the public internet, since when the API is unavailable, reconciliation fails entirely.

For the control plane loadbalancer, there is fall back code to enumerate all LBs and compare their properties. It is likely that this needs to be implemented in this case to make the feature usable in airgapped environments.

**What did you expect to happen:**

Cluster deletion continues.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-aws version: v2.0.2
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by wyike**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

Similar with https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/elb/loadbalancer.go#L1263, fallback to not using tagging client to filter resources in airgap env.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3987

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/cloud/services/gc/loadbalancer.go [hunk 1]
```diff
@@ -128,3 +130,150 @@ func (s *Service) deleteTargetGroup(ctx context.Context, targetGroupARN string)
...
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return arns, nil
+}
+
+// getProviderOwnedLoadBalancers gets cloud provider created LB(ELB) for this cluster, filtering by tag: kubernetes.io/cluster/<cluster-name>:owned.
+func (s *Service) getProviderOwnedLoadBalancers() ([]*AWSResource, error) {
```
Comment (Ankitasw) [inline pkg/cloud/services/gc/loadbalancer.go:183]: nit: `getProviderOwnedLoadBalancers` and `getProviderOwnedLoadBalancersV2` looks like almost duplicated functions, can we combine these two with the use of switch statements?

**Review by Ankitasw — COMMENTED** [review]
A small nit, otherwise LGTM

**Comment by Ankitasw** [comment]
@wyike thanks for making changes, did you try to test this with airgapped environment(or somehow simulate it), as our E2E tests doesn't test this scenario.

**Comment by wyike** [comment]
> @wyike thanks for making changes, did you try to test this with airgapped environment(or somehow simulate it), as our E2E tests doesn't test this scenario.

I tested in my local by:

```
       awsOutput, err := s.resourceTaggingClient.GetResourcesWithContext(ctx, &awsInput)
	if err != nil {
		// retry by listing all LB/TG/SG as using tagging client will fail in air-gapped environments
		s.scope.Error(err, "Getting resources with tagging client fails, fallback to query resources directly")
		return s.deleteResourcesWithoutTagging(ctx)
	}
```
changing `err!=nil` to `err==nil`.
1. create a basic unmanaged cluster
2. create several lb services (3 NLB, 3ELB service) using gc_workload.yaml 
3. Pasting one log on the gc deletion for an unmanaged cluster:
```
I0302 08:59:09.106289      92 logger.go:67] "deleting aws resources created by tenant cluster" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6"
E0302 08:59:09.526332      92 logger.go:83] "Getting resources with tagging client fails, fallback to query resources directly" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6"
I0302 08:59:13.776121      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:loadbalancer/net/a912890f0b408460b8a6ad55b90ebb15/c25525c98b4d7d5e Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb]}
I0302 08:59:13.776200      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:loadbalancer/net/ac6804d81b5d245e3bef5dccaa70f6fe/7b73aa33c91a0fa2 Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb-2]}
I0302 08:59:13.776244      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:loadbalancer/net/af18e6f52025848c19fa4b7b770cf2e3/0b703bda7b3fae80 Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb-1]}
I0302 08:59:13.776282      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:targetgroup/k8s-default-podinfon-98e4500c23/f6f7f455d0f1277a Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb-2]}
I0302 08:59:13.776313      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:targetgroup/k8s-default-podinfon-70a398276c/8ceab50be14d0916 Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb-1]}
I0302 08:59:13.776343      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:targetgroup/k8s-default-podinfon-6200382fb8/83cdeeff8c0a8491 Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb-1]}
I0302 08:59:13.776376      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:targetgroup/k8s-default-podinfon-909ae22197/0b4b4c95769326d2 Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb-2]}
I0302 08:59:13.776408      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:targetgroup/k8s-default-podinfon-ea5b8a5dc8/be26e4f3dc191f30 Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb]}
I0302 08:59:13.776456      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:elasticloadbalancing:us-east-2:566360789078:targetgroup/k8s-default-podinfon-25ffae083a/002aa9733c66c666 Tags:map[kubernetes.io/cluster/c6:owned kubernetes.io/service-name:default/podinfo-nlb]}
I0302 08:59:13.776695      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:ec2:fake-region:fake-account:security-group/sg-00c0b8c4d2c5e7fbd Tags:map[kubernetes.io/cluster/c6:owned]}
I0302 08:59:13.776730      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:ec2:fake-region:fake-account:security-group/sg-0f8085360fdce986e Tags:map[kubernetes.io/cluster/c6:owned]}
I0302 08:59:13.776759      92 logger.go:67] "Resource found:" controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="default/c6" namespace="default" name="c6" reconcileID=d412f9ca-8c12-4005-af4b-4611103d4afa cluster="default/c6" resource=&{ARN:arn:aws:ec2:fake-region:fake-account:security-group/sg-0a16b2deb943e988f Tags:map[kubernetes.io/cluster/c6:owned]}
```
     notes: 
       1.the elbs are deleted by pre-part https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/elb/loadbalancer.go#L562. If I comment this L562,  deleteResourcesWithoutTagging can also cover the deletion of elbs.
      2. all the lb/targetgroup/securitygroups will be deleted finally. Sometimes securitygroups deletion get dependency errors at first or second time deletions (I guess it's because not enough time after related LB deleted), but will be finally deleted after reconciles.

It's really a problem e2e tests cannot cover it if we use fallback mechanism.  I am also thinking that but have no good idea. What do you think? I am thinking if I can add one more annotation or whatever option so that user can set (optional) in future and accordingly in our code we can choose `deleteResources` or `deleteResourcesWithoutTagging` by if else (not interface way, interface cannot fallback). In this way, we can add one e2e test, and in production customer can just use fallback, ignoring to set any gc strategy.

**Comment by Ankitasw** [comment]
@wyike thanks for testing this locally. The results looks fine.

> Maybe I can open two more issues for following PRs: 1. option for user to choose gc strategy and add one e2e test 2. unit test for this strategy

1. I think in the original issue, @richardcase mentioned about strategy right? Are you referring to somewhat similar approach?
2.  If we still need strategy and new issue about it, what does this PR fulfill then?

**Comment by richardcase** [comment]
I was thinking that when you start CAPA you specify which GC strategy you want to use (with the default being resource tagging). And not falling back on error to a different mechanism.

@Skarlso @Ankitasw - keen to get your thoughts on fallback vs explicitly specifying GC strategy

**Comment by wyike** [comment]
The original gc strategy code is also in this comment https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3987#issuecomment-1435513103 , avaliable to review.

Hi @richardcase  do you have idea how to specify the gc strategy at the begging by customers?

**Comment by richardcase** [comment]
> Hi @richardcase do you have good idea how to specify the gc strategy at the beginning by customers?

I was thinking a flag. And the flags value is used to create an instance of strategy implementation.

**Comment by wyike** [comment]
Close this PR for now since we decide to expose option to let user choose different gc strategy.

**Pull request closed without merging**